File: Examples/fitUncoatedParticles.py
import random
import numpy as np
import ACMHystAreaBrown
import sys

# Generate random numbers (similar to generateRandomDoubleb)
def generate_random_double(first, last):
    return random.uniform(first, last)

# Read data from the file (similar to readAreas)
def read_areas(filename):
    mu0 = 4 * np.pi * 1e-7 * 1e9  # Magnetic permeability constant
    fb = []
    areas = []

    with open(filename, 'r') as infile:
        for line in infile:
            # Ignore empty lines or comments
            if not line.strip() or line.startswith('#'):
                continue

            # Parse the line
            parts = line.split()
            if len(parts) < 3:
                raise ValueError(f"Error reading line: {line.strip()}")
            freq, field, area = map(float, parts)

            # Create field parameters and adjust area
            fp = ACMHystAreaBrown.FieldParameters(amplitude=field * 1e-6 * mu0, frequency=freq * 1e-6)
            fb.append(fp)
            areas.append(area / field)

    return fb, areas

# Generate initial guesses (similar to generateInitialGuesses)
def generate_initial_guesses(n_temperatures):
    initial_guesses = []
    for _ in range(n_temperatures):
        rc = generate_random_double(5, 50)
        msat = generate_random_double(0.00005, 0.0005)
        num_particles = generate_random_double(0.01, 10)

        guess = {
            "coreRadius": rc,
            "msat": msat,
            "numParticles": num_particles,
        }
        initial_guesses.append(guess)
    return initial_guesses

# Configure parameters
def main():
    # ParallelTempering configuration
    mc_temperatures = [1e-6, 1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1]
    jump_size       = [0.1] * len(mc_temperatures)
    initial_guesses = generate_initial_guesses(len(mc_temperatures))

    pt_par = ACMHystAreaBrown.PTParameters()
    pt_par.temperatures   = mc_temperatures
    pt_par.numStepsFinish = 2000
    pt_par.numStepsSwap   = 100
    pt_par.maxIterations  = 2000
    pt_par.jumpSize       = jump_size
    pt_par.tolerance      = 0.00001
    pt_par.printSteps     = 10000

    # GaussNewton configuration
    gn_par = ACMHystAreaBrown.GNParameters()
    gn_par.maxIterations  = 500
    gn_par.tolerance      = 1e-4
    gn_par.regularization = 1e-6
    gn_par.printSteps     = 500

    # Read areas from a file
    
    filename  = "Areas_rc_15.data"
    fb, areas = read_areas(filename)
    
    # Additional parameters configuration
    extra_parameters = {
        "kBT": 0.0041124,
        "viscosity": 0.0009,
        "coatingWidth": 0.0,
    }

    target_parameters = {"coreRadius": 15,
                         "msat":0.000150,
                         "numParticles":1};

    # Fit the areas using the Python wrapper
    fit_result = ACMHystAreaBrown.fitAreas(
        fb, areas, pt_par, gn_par, initial_guesses, extra_parameters
    )

    # Print adjusted parameters
    print("Adjusted Parameters:")
    for key, value in fit_result[0].items():  # fit_result[0] contains the adjusted parameters
        print(f"{key}: {value}")

    print("\nTarget Parameters:")
    for key, value in target_parameters.items():
        print(f"{key}: {value}")

    # Optionally, print the adjusted errors
    print("\nStandard Errors:")
    for key, value in fit_result[1].items():  # fit_result[1] contains the adjusted errors
        print(f"{key}: {value}")


if __name__ == "__main__":
    main()

File: Examples/fitUncoatedParticles.cpp
#include<random>
#include"FitMagneticArea.h"

inline std::mt19937& getRngb() {
  static std::random_device rd;
  static std::mt19937 rng(rd());
  return rng;
}

inline double generateRandomDoubleb(double first, double last) {
  std::uniform_real_distribution<double> dist(first, last);
  return dist(getRngb());
}

void readAreas(const std::string& filename, std::vector<FieldParameters>& fb, std::vector<double>& areas) {

  double mu0       = 4*M_PI*1e-7*1e9;
  std::ifstream infile(filename);
  if (!infile) {
    throw std::runtime_error("Could not open file " + filename);
  }
  
  std::string line;
  while (std::getline(infile, line)) {
    // Ignore comment lines starting with #
    if (line.empty() || line[0] == '#') {
      continue;
    }
    
    std::istringstream iss(line);
    double freq, field, area;
    if (!(iss >> freq >> field >> area)) {
      throw std::runtime_error("Error reading line: " + line);
    }
    
    FieldParameters fp; fp.freq = freq*1e-6; fp.field = field*1e-6*mu0;
    fb.push_back(fp);
    areas.push_back(area/(field));
  }
  
  infile.close();
}

std::vector<std::map<std::string, double>> generateInitialGuesses(int nTemperatures){

  //Be careful, here the standard deviation is in units of rc: std = rc*std_in
  std::vector<std::map<std::string, double>> initialGuesses(nTemperatures);
  auto now    = std::chrono::system_clock::now();
  auto now_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count();
  
  fori(0, nTemperatures){
    double rc            = generateRandomDoubleb(5,50);
    double msat          = generateRandomDoubleb(0.00005,0.0005);
    double rcstd         = generateRandomDoubleb(0.05,1);
    double numParticles  = generateRandomDoubleb(0.01,10);
    std::map<std::string, double> initialGuess = {
      {"coreRadius", rc},
      {"msat", msat},
      {"numParticles", numParticles},
    };
    initialGuesses[i] = initialGuess;    
  }
  return initialGuesses;
}




int main(int argc, char* argv[]){
  
  std::vector<double> MCtemperatures = {1e-6, 1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1};
  std::vector<double> jumpSize(MCtemperatures.size(), 0.1);
  auto initialGuesses = generateInitialGuesses(MCtemperatures.size());

  ParallelTempering::Parameters ptPar;
  ptPar.temperatures   = MCtemperatures;
  ptPar.numStepsFinish = 2000;
  ptPar.numStepsSwap   = 100;
  ptPar.maxIterations  = 2000;
  ptPar.jumpSize       = jumpSize;
  ptPar.tolerance      = 0.00001;
  ptPar.numThreads     = 5;
  ptPar.printSteps     = 1000;
  
  
  GaussNewton::Parameters gnPar;
  gnPar.maxIterations     = 500;
  gnPar.tolerance         = 1e-4;
  gnPar.regularization    = 1e-6;
  gnPar.printSteps        = 5;
  
  std::string filename = std::string(argv[1]);
  std::vector<FieldParameters> fb;
  std::vector<double> areas;
  readAreas(filename, fb, areas);

  std::map<std::string, double> extraParameters = {
    {"kBT", 0.0041124},
    {"viscosity", 0.0009},
    {"coatingWidth", 0.0},
  };
  
  auto params = fitAreas(fb, areas, ptPar, gnPar,
                         initialGuesses, extraParameters);
  
  for (const auto& pair : params.parameters) {
    std::cout << pair.first << ": " << pair.second << std::endl;
  }
}
